[VarDumper] Adds support for ReflectionUnionType to VarDumper

michaeldnelson opened the pull request:
Fixes a bug with VarDumper when dumping a ReflectionUnionType.
> PHP Error:  Call to undefined method ReflectionUnionType::isBuiltin()

Notes:
* One of the existing tests relies on its position in the test file. I had to modify its expected line number.
* There is an existing trailing space around line 367 in an expected value.
  I'm not sure if this was left for BC reasons but it seems like a bug if the dumper is leaving trailing spaces.

| Q             | A
| ------------- | ---
| Branch?       | 4.4
| Bug fix?      | yes
| New feature?  | no
| Deprecations? | no
| Tickets       | 
| License       | MIT
| Doc PR        | 

This commit fixes a crash when dumping ReflectionUnionType.  The code is minimal but uses a few extra lines to preserve key order for bc and consistency.  Additionally, there is an else condition that is currently unreachable but is defensive should they add additional subtypes of ReflectionType. Tests are included.  Please let me know if you have any questions or suggestions.  Thanks for Symfony it's a wonderful project.

mndevel committed:
[VarDumper] Adds support for ReflectionUnionType to VarDumper

michaeldnelson commented:
Looking at ci It makes sense **tests** would break using php < 8.  I'm not quite sure the best way to add this without breaking and I don't have multiple php versions running at the moment.  I'm happy to take and suggestions about how to test this feature without breaking php < 8. Two possibilities I can think of are multiple classes or a single class with a conditional as below, either used in conjunction with the @require php8 annotation.  I'm not sure if the parser lints a file on include, if that is the case I have no idea how to resolve this.

ReflectionPropertyFixture.php
```
if (\PHP_VERSION_ID >= 80000) {
    class ReflectionPropertyFixture
    {
        public int $a;
        public int|string $b;
    }
} else {
    class ReflectionPropertyFixture
    {
        public int $a;
    }
}
```

derrabus commented on src/Symfony/Component/VarDumper/Tests/Caster/ReflectionCasterTest.php:
```diff
@@ -314,6 +373,7 @@ class: "Symfony\Component\VarDumper\Tests\Caster\ReflectionCasterTest"
 
     public function testGenerator()
     {
+
```
Should be reverted

derrabus commented:
Thank you for working on this bugfix.

> This commit fixes a crash when dumping ReflectionUnionType. This targets 5.2 due to it being a bugfix that only impacts PHP 8.

Your pull request targets 5.x right now which is probably mot what you wanted. If you can reproduce the crash on 4.4 as well (and I’m pretty sure that this is the case), then we should target 4.4 instead.

> Thanks for Symfony it's a wonderful project.

❤️

derrabus commented on src/Symfony/Component/VarDumper/Tests/Caster/ReflectionCasterTest.php:
```diff
@@ -227,6 +228,64 @@ public function testReflectionParameterNullableUnion()
         );
     }
 
+    public function testReflectionPropertyScalar()
```
This test will fail on PHP 7 because your new fixture class cannot be parsed by a PHP 7 interpreter. Either flag this test as `@requires PHP 8` or create a separate fixture for PHP 7.

michaeldnelson replied:
Yup, I noticed that in CI, I will make this change.

derrabus commented on src/Symfony/Component/VarDumper/Tests/Caster/ReflectionCasterTest.php:
```diff
@@ -227,6 +228,64 @@ public function testReflectionParameterNullableUnion()
+        );
+    }
+
+    public function testReflectionNamedType()
```
Same here.

michaeldnelson replied:
ditto.

michaeldnelson commented:
@derrabus I've rebased onto 4.4 and resolved the php7 issues as suggested.  CI is stuck due to a perceived conflict before I switched the target branch of the PR so I am unable to verify it is passing.  Please let me know if there are any further issues, and thanks again.

derrabus commented on src/Symfony/Component/VarDumper/Caster/ReflectionCaster.php:
```diff
@@ -377,7 +387,7 @@ private static function addExtra(array &$a, \Reflector $c)
     }
 
-    private static function addMap(array &$a, \Reflector $c, array $map, string $prefix = Caster::PREFIX_VIRTUAL)
+    private static function addMap(array &$a, object $c, array $map, string $prefix = Caster::PREFIX_VIRTUAL)
```
Symfony 4.4 supports PHP 7.1, but the `object` type hint was introduced with PHP 7.2.

michaeldnelson replied:
There is an option to check for is_object, but given that this is a private function on a bugfix branch, and we know it is type hinted on the 5.x branch it seemed more likely to cause confusion and/or cause merge complications than provide a useful error.  Removal seemed like the best option.

stof commented on src/Symfony/Component/VarDumper/Caster/ReflectionCaster.php:
```diff
@@ -96,11 +96,21 @@ public static function castType(\ReflectionType $c, array $a, Stub $stub, $isNes
+            ];
+        } else {
+            // As of PHP 8.0 there are only two children of ReflectionType (abstract) so this is unreachable.
+            $a[$prefix.'allowsNull'] = $c->allowsNull();
```
should we rather throw an `UnexpectedValueException` in this currently unreadable branch, so that we get a proper reporting if it becomes reachable ?

stof replied:
another option (potentially cleaner) is to split casters for ReflectionUnionType and ReflectionNamedType

michaeldnelson replied:
> should we rather throw an UnexpectedValueException in this currently unreadable branch, so that we get a proper reporting if it becomes reachable ?

I'm happy to update with the exception, if that is preferable.  I'm not sure what the Symfony change management process is like so I'm happy to go with your suggestion here if you think that is a better fit.

> another option (potentially cleaner) is to split casters for ReflectionUnionType and ReflectionNamedType

The logic for handling the abstract ReflectionType is minimal, if we split it into two functions then everywhere someone calls this function they would need to check type before calling.  If the return types were significantly different that could be preferable, but this is building a fairly complex array, almost like a view, so I'm unclear on the benefit. 

We could possibly then create a third function that wraps the other two and does the type checking proxy but, to me, that seems like it would be harder to read if the differences remain small.

derrabus replied:
> The logic for handling the abstract ReflectionType is minimal, if we split it into two functions then everywhere someone calls this function they would need to check type before calling.

That change would be too heavy for a bugfix, I think. I'm fine with the `if`, but I agree that we should throw if the unreachable `else` is actually hit.

michaeldnelson replied:
Apologies, I know my comment here led us down this particular path, but I was just thinking about how I would test this and I realized it is actually possible.

```
class FakeReflection extends \ReflectionType {
    public function allowsNull() {
        return false;
    }

    public function __toString()
    {
       return 'fake';
    }
};
$t = new FakeReflection();
echo match(true) {
    $t instanceof \ReflectionNamedType => 'named',
    $t instanceof \ReflectionUnionType => 'union',
    $t instanceof \ReflectionType => 'abstract',
    default => 'error'
};
```
```> 'abstract'```

michaeldnelson replied:
I need to remove this comment.

We could add isBuiltin() and a string cast to to the final else for php versions <8 to remove the possibility of a bc break.

For php versions > 8 we should not add these fields.   isBuiltin() does not exist on the abstract in php 8 and toString() has been depecrated (though functional) since 7.1.  This is not a bc break for 8 because this case never worked on 8 due to isBuiltin() causing a crash.

michaeldnelson replied:
I'm going to mark this resolved as it was based on a misconception that this is unreachable.

nicolas-grekas commented on src/Symfony/Component/VarDumper/Tests/Fixtures/ExtendsReflectionTypeFixture.php:
```diff
@@ -0,0 +1,22 @@
+<?php
+
+
```
extra blank line should be removed (same below)

michaeldnelson replied:
Done

nicolas-grekas commented on src/Symfony/Component/VarDumper/Tests/Fixtures/ExtendsReflectionTypeFixture.php:
```diff
@@ -0,0 +1,22 @@
+
+class ExtendsReflectionTypeFixture extends \ReflectionType
+{
+    public function allowsNull()
```
let's add all possible return types on methods

nicolas-grekas commented on src/Symfony/Component/VarDumper/Caster/ReflectionCaster.php:
```diff
@@ -96,11 +96,21 @@ public static function castType(\ReflectionType $c, array $a, Stub $stub, $isNes
+                'types' => 'getTypes',
+            ]);
+        } else {
+            //Userland may extend reflection type.
```
can be removed (the test is enough a clue if one wonders why this code path)

michaeldnelson replied:
Done

nicolas-grekas reviewed (approved):
LGTM with minor comments, thanks.

derrabus reviewed (approved):
👍🏻 (after Nicolas' comments have been addressed)

derrabus commented on src/Symfony/Component/VarDumper/Tests/Fixtures/ExtendsReflectionTypeFixture.php:
```diff
@@ -0,0 +1,21 @@
+
+class ExtendsReflectionTypeFixture extends \ReflectionType
+{
+    public function allowsNull() : bool
+    {
+        return false;
+    }
+
+    public function isBuiltin() : bool
+    {
+        return false;
+    }
+
+    public function __toString() : string
+    {
+        return 'fake';
+    }
```
```suggestion
    public function allowsNull(): bool
    {
        return false;
    }

    public function isBuiltin(): bool
    {
        return false;
    }

    public function __toString(): string
    {
        return 'fake';
    }
```

michaeldnelson replied:
Oops, I didn't realize the spaces were out of style.

michaeldnelson replied:
Thanks!

fabpot commented:
Thank you @michaeldnelson.
